src/algorithms/tonal/multipitch.cpp

```diff
@@ -0,0 +1,406 @@
+
+  vector<Real> frameWindowed;
+  _windowing->input("frame").set(frame);
+  _windowing->output("frame").set(frameWindowed);
+
+  // Spectral peaks
+  vector<Real> frameSpectrum;
+  _spectrum->input("frame").set(frameWindowed);
+  _spectrum->output("spectrum").set(frameSpectrum);
+
+  vector<Real> frameFrequencies;
+  vector<Real> frameMagnitudes;
+  _spectralPeaks->input("spectrum").set(frameSpectrum);
+  _spectralPeaks->output("frequencies").set(frameFrequencies);
+  _spectralPeaks->output("magnitudes").set(frameMagnitudes);
+
+  // Spectral whitening
+  vector<Real> frameWhiteMagnitudes;
+  _spectralWhitening->input("spectrum").set(frameSpectrum);
+  _spectralWhitening->input("frequencies").set(frameFrequencies);
+  _spectralWhitening->input("magnitudes").set(frameMagnitudes);
+  _spectralWhitening->output("magnitudes").set(frameWhiteMagnitudes);
+    
+  // Pitch salience contours
+  vector<Real> frameSalience;
+  _pitchSalienceFunction->input("frequencies").set(frameFrequencies);
+  _pitchSalienceFunction->input("magnitudes").set(frameMagnitudes);
+  _pitchSalienceFunction->output("salienceFunction").set(frameSalience);
+
+  vector<Real> frameSalienceBins;
+  vector<Real> frameSalienceValues;
+  _pitchSalienceFunctionPeaks->input("salienceFunction").set(frameSalience);
+  _pitchSalienceFunctionPeaks->output("salienceBins").set(frameSalienceBins);
+  _pitchSalienceFunctionPeaks->output("salienceValues").set(frameSalienceValues);
+
+
+  vector<vector<Real> > peakBins;
+  vector<vector<Real> > peakSaliences;
+
+  vector<Real> nearestBinWeights;
+  nearestBinWeights.resize(binsInSemitone + 1);
+  for (int b=0; b <= binsInSemitone; b++) {
+    nearestBinWeights[b] = pow(cos((Real(b)/binsInSemitone)* M_PI/2), 2);
+  }
+    
+  vector<Real> harmonicWeights;
+  harmonicWeights.clear();
+  harmonicWeights.reserve(numberHarmonicsMax);
+  for (int h=0; h<numberHarmonicsMax; h++) {
+    harmonicWeights.push_back(pow(0.8, h));
+  }
+    
+  while (true) {
+    // get a frame
+    _frameCutter->compute();
+
+    if (!frame.size()) {
+      break;
+    }
+
+    _windowing->compute();
+
+    // calculate spectrum
+    _spectrum->compute();
+
+    // calculate spectral peaks
+    _spectralPeaks->compute();
+      
+    // whiten the spectrum
+      _spectralWhitening->compute();
+
+    // calculate salience function
+    _pitchSalienceFunction->compute();
+
+    // calculate peaks of salience function
+    _pitchSalienceFunctionPeaks->compute();
+      
+    // no peaks in this frame
+      if (!frameSalienceBins.size()){
+          continue;
+      }
+    ///////////////////////////////////////////////////////////////////////
+    // Joint F0 estimation (pitch salience function peaks as candidates) //
+    ///////////////////////////////////////////////////////////////////////
+      
+    // compute the cent-scaled spectrum
+    vector<float> centSpectrum;
+    int numberBins=frequencyToCentBin(sampleRate/2);
+    centSpectrum.resize(numberBins);
```

Comment:
Optimize: Computing numberBins and resizing centSpectrum should be in configure().